servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutHttpServiceFilter.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.utils;
+
+import io.servicetalk.concurrent.api.Executor;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpExecutionStrategyInfluencer;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.http.utils.TimeoutHttpRequesterFilter.simpleDurationTimeout;
+
+/**
+ * A {@link StreamingHttpServiceFilter} that adds support for request/response timeouts.
+ *
+ * <p>The order with which this filter is applied may be highly significant. For example, appending it before a retry
+ * filter would have different results than applying it after the retry filter; timeout would apply for all retries vs
+ * timeout per retry.
+ */
+public final class TimeoutHttpServiceFilter
+        implements StreamingHttpServiceFilterFactory, HttpExecutionStrategyInfluencer {
+
+    private final TimeoutHttpRequesterFilter.TimeoutFromRequest timeoutForRequest;
+
+    @Nullable
+    private final Executor timeoutExecutor;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param duration the timeout {@link Duration}
+     */
+    public TimeoutHttpServiceFilter(Duration duration) {
+        this(simpleDurationTimeout(duration));
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param duration the timeout {@link Duration}
+     * @param timeoutExecutor the {@link Executor} to use for managing the timer notifications
+     */
+    public TimeoutHttpServiceFilter(Duration duration,
+                                    Executor timeoutExecutor) {
```

Comment:
Can we make client/server filters consistent and also add an overload with `fullRequestResponse`? Sounds like a useful feature for the server-side as well to decide where timeout should be applied.

Reply:
Any thoughts here?

Reply:
I have added fullRequestResponse support to the Service filter as well and the implementations are even more similar than before. I haven't tried to share the code between them yet as they implement slightly different interfaces and get the response in slightly different ways.